statsmodels/tsa/stl/mstl.py

```diff
@@ -0,0 +1,245 @@
+        Must be an odd integer, and should normally be >= 7 (default). If None
+        then default values determined using 7 + 4 * np.arange(1, n + 1, 1)
+        where n is number of seasonal components.
+    lmbda : {float, str, None}, optional
+        The lambda parameter for the Box-Cox transform to be applied to `endog`
+        prior to decomposition. If None, no transform is applied. If "auto", a
+        value will be estimated that maximizes the log-likelihood function.
+    iterate : int, optional
+        Number of iterations to use to refine the seasonal component.
+    stl_kwargs: dict, optional
+        Arguments to pass to STL.
+
+    See Also
+    --------
+    statsmodels.tsa.seasonal.STL
+
+    References
+    ----------
+    .. [1] K. Bandura, R.J. Hyndman, and C. Bergmeir (2021)
+    MSTL: A Seasonal-Trend Decomposition Algorithm for Time Series with
+    Multiple Seasonal Patterns. arXiv preprint arXiv:2107.13462.
+
+    Examples
+    --------
+    Start by creating a toy dataset with multiple seasonal components.
+
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+    >>> register_matplotlib_converters()
+    >>> t = np.arange(1, 1000)
+    >>> trend = 0.0001 * t ** 2 + 100
+    >>> daily_seasonality = 5 * np.sin(2 * np.pi * t / 24)
+    >>> weekly_seasonality = 10 * np.sin(2 * np.pi * t / (24 * 7))
+    >>> noise = np.random.randn(len(t))
+    >>> y = trend + daily_seasonality + weekly_seasonality + noise
+
+    Use MSTL to decompose the time series into multiple seasonal components
+    with periods 24 (daily) and 24*7 (weekly).
+
+    >>> from statsmodels.tsa.seasonal import MSTL
+    >>> res = MSTL(data, periods=(24, 24*7)).fit()
+    >>> res.plot()
+    >>> plt.show()
+    """
+
+    def __init__(
+        self,
+        endog: ArrayLike1D,
+        *,
+        periods: Optional[Union[int, Sequence[int]]] = None,
+        windows: Optional[Union[int, Sequence[int]]] = None,
+        lmbda: Optional[Union[float, str]] = None,
+        iterate: int = 2,
+        stl_kwargs: Optional[Dict[str, Union[int, bool, None]]] = None,
+    ):
+        self.endog = endog
+        self._y = self._to_1d_array(endog)
+        self.nobs = self._y.shape[0]
+        self.lmbda = lmbda
+        self.periods = self._process_periods(periods)
+        self.windows = self._process_windows(windows)
+        self.iterate = iterate
+        self._stl_kwargs = self._remove_overloaded_stl_kwargs(
+            stl_kwargs if stl_kwargs else {}
+        )
+
+        if len(self.periods) != len(self.windows):
+            raise ValueError("Periods and windows must have same length")
+
+    def fit(self):
+        """
+        Estimate multiple season components as well as trend and residuals
+        components.
+
+        Returns
+        -------
+        DecomposeResult
+            Estimation results.
+        """
+        periods, windows = self._sort_periods_and_windows()
```

Comment:
Somethign is wrong here. You first get `periods` and `windows`, and then right below you access `self.preiods`.  Should you be using these sorted periods below?  If so, could you add a test where you have periods out-of-order and also inorder and ensure the results are the same?

Reply:
If you really need to sort, should you not just do this in the `__init__` so that there is a consistent canonical representation of models that have the same periods/windows in any order?

Reply:
The sorting logic is part of the algorithm which is why I was tempted to keep it in `.fit()`. But I think your suggestion also makes sense so I've refactored the logic for handling the various input types and the sorting of `windows` and `periods`, moved it to the `__init__`, and implemented  the suggested new test.